README.md
# TFC-Operator
TFC-Operator is the Kubenretes Native Operator with GitOPs of Terraform. Supports Github/Gitlab (Public/Private), and uses HCL (Hashicorp Configuration Language) to provision resources in Cloud Platform (e.g. AWS, Azure, GCP, vSphere.. etc)

# Quick Start
This guide contains information on creating an EC2 instance on AWS through Terraform Claims (CR).

## 1. Public Repo
(1) Push the HCL (Hashicorp Configuration Language) code that defines Terraform resources to the git repository (Github / GitLab)

### Example.
```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.27"
    }
  }

  required_version = ">= 0.14.9"
}

provider "aws" {
  access_key = "<AWS_ACCESS_KEY_ID>"
  secret_key = "<AWS_SECRET_ACCESS_KEY>"
  region = "us-east-1"
}

resource "aws_instance" "app_server" {
  ami           = "ami-0003076ab1664cd23"
  instance_type = "t2.micro"

  tags = {
    Name = "ExampleAppServerInstance"
  }
}
```

(2) Create a Terraform claim based on the previously created Git Repo information
```bash
kubectl apply -f samples/public/01_claim.yaml
```

## 2. Private Repo
(1) Push the HCL (Hashicorp Configuration Language) code that defines Terraform resources to the git repository (Github / GitLab) 
### Example. (same as 1-(1))
```hcl
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.27"
    }
  }

  required_version = ">= 0.14.9"
}

provider "aws" {
  access_key = "<AWS_ACCESS_KEY_ID>"
  secret_key = "<AWS_SECRET_ACCESS_KEY>"
  region = "us-east-1"
}

resource "aws_instance" "app_server" {
  ami           = "ami-0003076ab1664cd23"
  instance_type = "t2.micro"

  tags = {
    Name = "ExampleAppServerInstance"
  }
}
```

(2) Create a secret holding an access token that can access the private repo
```bash
kubectl apply -f samples/private/01_secret.yaml
```
(3) Create a Terraform claim based on the previously created Git Repo information
```bash
kubectl apply -f samples/private/02_claim.yaml
```



CHANGELOG.md
# tfc-operator changelog!!
All notable changes to this project will be documented in this file.

<!-------------------- v5.0.36.3 start -------------------->

## tfc-operator_5.0.36.3 (2022. 12. 06. (화) 17:31:02 KST)

### Added
  - [feat] add generating v1beta1 manifeste to makefile by sjoh0704

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.36.3 end --------------------->

<!-------------------- v5.0.36.2 start -------------------->

## tfc-operator_5.0.36.2 (2022. 11. 18. (금) 10:38:10 KST)

### Added

### Changed
  - [mod] crd description 재변경 by sjoh0704

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.36.2 end --------------------->

<!-------------------- v5.0.36.1 start -------------------->

## tfc-operator_5.0.36.1 (2022. 11. 18. (금) 10:16:18 KST)

### Added
  - [feat] fix crd manifest error by sjoh0704

### Changed

### Fixed
  - [ims][294005] 테라폼 클레임 crd description 수정 요청 by sjoh0704

### CRD yaml

### Etc

<!--------------------- v5.0.36.1 end --------------------->

<!-------------------- v5.0.36.0 start -------------------->

## tfc-operator_5.0.36.0 (2022. 11. 14. (월) 18:48:25 KST)

### Added
  - [feat] 기능 고도화, 불필요한 requeue 제거, reject error 해결, tf 변수 사용법 간소화 by sjoh0704

### Changed
  - [mod] terraform 변수 입력시 stuck 걸리는 error fix by sjoh0704
  - [mod] error 수정 by sjoh0704

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.36.0 end --------------------->

<!-------------------- v5.0.35.0 start -------------------->

## tfc-operator_5.0.35.0 (2022. 11. 11. (금) 11:30:55 KST)

### Added
  - [feat] applied 상태에서 tfapplyclaim을 지우지 못하도록 웹훅 추가 by sjoh0704
  - [feat] TFC_WORKER 환경변수 check logic 추가, crd description 수정 by sjoh0704
  - [feat] 웹훅 추가, printcolumn 추가, shorts 추가 by sjoh0704

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.35.0 end --------------------->

<!-------------------- v5.0.34.1 start -------------------->

## tfc-operator_5.0.34.1 (2022. 11. 03. (목) 21:17:47 KST)

### Added
  - [feat] git clone시 gitlab, github 분기로직 추가 by sjoh0704

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.34.1 end --------------------->

<!-------------------- v5.0.34.0 start -------------------->

## tfc-operator_5.0.34.0 (2022. 11. 02. (수) 16:20:07 KST)

### Added

### Changed
  - [mod] 기본 setting 수정 by sjoh0704

### Fixed

### CRD yaml

### Etc
  - [etc] Kubernetes API Call 테스트 스크립트 작성 by gyeongyeol-choi

<!--------------------- v5.0.34.0 end --------------------->

<!-------------------- v5.0.29.0 start -------------------->

## tfc-operator_5.0.29.0 (2022. 07. 12. (화) 13:25:21 KST)

### Added
  - [feat] 클라우드 리소스 추가 시 (terraform apply), 클레임 리소스 삭제 방지를 위한 Finalizer 적용 by gyeongyeol choi

### Changed

### Fixed

### CRD yaml

### Etc
  - [etc] 에러 코드 삭제 by gyeongyeol-choi
  - [etc] 불필요한 주석 제거 by gyeongyeol-choi

<!--------------------- v5.0.29.0 end --------------------->

<!-------------------- v5.0.25.1 start -------------------->

## tfc-operator_5.0.25.1 (2022. 05. 18. (수) 19:18:19 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc
  - [etc] k8s 1.22 업그레이드에 따른 CRD 업데이트 by gyeongyeol-choi
  - [etc] k8s 1.22 버전 업그레이드에 따른 apiVersion 업그레이드 by gyeongyeol-choi
  - [etc] TFC-Operator 사용법 가이드 작성 by GitHub
  - [etc] 테스트용 샘플 매니페스트 파일 업로드 by gyeongyeol-choi

<!--------------------- v5.0.25.1 end --------------------->

<!-------------------- v5.0.25.0 start -------------------->

## tfc-operator_5.0.25.0 (2021. 08. 27. (금) 18:58:39 KST)

### Added

### Changed
  - [mod] Private Git 인증처리 로직 수정 (ID/PW -> Personal Access Token) by gyeongyeol-choi

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.25.0 end --------------------->

<!-------------------- v5.0.24.0 start -------------------->

## tfc-operator_5.0.24.0 (2021. 08. 19. (목) 17:04:34 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.24.0 end --------------------->

<!-------------------- v5.0.23.0 start -------------------->

## tfc-operator_5.0.23.0 (2021. 08. 12. (목) 13:07:38 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.23.0 end --------------------->

<!-------------------- v5.0.22.0 start -------------------->

## tfc-operator_5.0.22.0 (2021. 08. 05. (목) 15:18:09 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.22.0 end --------------------->

<!-------------------- v5.0.21.0 start -------------------->

## tfc-operator_5.0.21.0 (2021. 07. 29. (목) 15:36:02 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc
  - [etc] 코드 리팩토링 (전역변수 처리, 공통로직 함수 모듈화 등) by gyeongyeol-choi

<!--------------------- v5.0.21.0 end --------------------->

<!-------------------- v5.0.20.0 start -------------------->

## tfc-operator_5.0.20.0 (2021. 07. 22. (목) 14:21:22 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.20.0 end --------------------->

<!-------------------- v5.0.19.0 start -------------------->

## tfc-operator_5.0.19.0 (2021. 07. 15. (목) 17:23:08 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.19.0 end --------------------->

<!-------------------- v5.0.18.0 start -------------------->

## tfc-operator_5.0.18.0 (2021. 07. 09. (금) 09:40:27 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc

<!--------------------- v5.0.18.0 end --------------------->

<!-------------------- v5.0.17.7 start -------------------->

## tfc-operator_5.0.17.7 (2021. 07. 08. (목) 18:09:31 KST)

### Added

### Changed

### Fixed

### CRD yaml

### Etc
  - [etc] Terraform Worker Pod 생성 지연 시간 변경 (5 sec -> 15 sec) by gyeongyeol-choi
  - [etc] CRD 업데이트에 따른 key mapping 파일 수정 by gyeongyeol-choi

<!--------------------- v5.0.17.7 end --------------------->

<!-------------------- v5.0.17.6 start -------------------->

## tfc-operator_5.0.17.6 (2021. 07. 08. (목) 13:33:28 KST)

### Added

### Changed

util/terraform.go
package util

import (
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/prometheus/common/log"

	v1 "k8s.io/api/core/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/kubernetes/scheme"
	restclient "k8s.io/client-go/rest"
	"k8s.io/client-go/tools/remotecommand"

	claimv1alpha1 "github.com/tmax-cloud/tfc-operator/api/v1alpha1"
)

func ExecClone(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	var protocol string

	if strings.Contains(apply.Spec.URL, "http://") {
		protocol = "http://"
	} else if strings.Contains(apply.Spec.URL, "https://") {
		protocol = "https://"
	}

	url := strings.TrimLeft(apply.Spec.URL, protocol)

	cmd := ""
	//github인 경우와 gitlab인 경우 token url 순서가 다름
	if strings.Contains(apply.Spec.URL, "github") {
		if apply.Spec.Type == "private" {
			url = protocol + "$GIT_TOKEN:x-oauth-basic" + "@" + url
		} else {
			url = protocol + "TMP_TOKEN:x-oauth-basic" + "@" + url
		}
		rep_token := "GIT_TOKEN=$(echo $GIT_TOKEN | sed -e 's/\\!/%21/g' -e 's/\\#/%23/g' -e 's/\\$/%24/g' -e 's/\\&/%26/g' -e \"s/'/%27/g\" -e 's/(/%28/g' -e 's/)/%29/g' -e 's/\\*/%2A/g'  -e 's/\\+/%2B/g' -e 's/\\,/%2C/g' -e 's/\\//%2F/g' -e 's/\\:/%3A/g' -e 's/\\;/%3B/g' -e 's/\\=/%3D/g' -e 's/\\?/%3F/g' -e 's/\\@/%40/g' -e 's/\\[/%5B/g'  -e 's/\\]/%5D/g');"
		cmd = rep_token + "git clone " + url + " " + HCL_DIR

	} else {
		if apply.Spec.Type == "private" {
			url = protocol + "x-oauth-basic:$GIT_TOKEN" + "@" + url
		} else {
			url = protocol + "x-oauth-basic:TMP_TOKEN" + "@" + url
		}
		cmd = "git clone " + url + " " + HCL_DIR

	}

	// SSL 미인증 VCS (e.g. Gitlab)을 위한 로직 처리
	cmd = "git config --global http.sslVerify false;" + cmd

	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecBranchCheckout(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	cmd := "cd " + HCL_DIR + ";" + "git checkout -t origin/" + apply.Spec.Branch
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecTerraformDownload(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	version := apply.Spec.Version

	cmd := "cd /tmp;" +
		fmt.Sprintf("wget %s/%s/%s_%s_linux_amd64.zip;", TERRAFORM_BINARY_URL, version, TERRAFORM_BINARY_NAME, version) +
		fmt.Sprintf("wget %s/%s/%s_%s_SHA256SUMS;", TERRAFORM_BINARY_URL, version, TERRAFORM_BINARY_NAME, version) +
		fmt.Sprintf("unzip -o -d /bin %s_%s_linux_amd64.zip;", TERRAFORM_BINARY_NAME, version) +
		"rm -rf /tmp/build;"

	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecTerraformInit(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	var cmd string

	versions := strings.Split(apply.Spec.Version, ".")

	majorVersion, _ := strconv.Atoi(versions[0])
	minorVersion, _ := strconv.Atoi(versions[1])

	if int(majorVersion) >= 1 || minorVersion >= 15 {
		cmd = "cd " + HCL_DIR + ";" + "terraform init"
	} else {
		cmd = "cd " + HCL_DIR + ";" + "terraform init -verify-plugins=false"
	}

	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecGitPull(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	cmd := "cd " + HCL_DIR + ";" + "git pull"
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecGetCommitID(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	cmd := "cd " + HCL_DIR + ";" +
		"git log --pretty=format:\"%H\" | head -n 1"

	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecTerraformPlan(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	variables := GetTerraformVariables(apply)
	cmd := "cd " + HCL_DIR + ";" + "terraform plan" + variables
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecTerraformApply(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	variables := GetTerraformVariables(apply)
	cmd := "cd " + HCL_DIR + ";" + "terraform apply -auto-approve" + variables
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecReadState(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	cmd := "cd " + HCL_DIR + ";" +
		"cat terraform.tfstate"
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecRevertCommit(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	cmd := "cd " + HCL_DIR + ";" +
		"git reset " + apply.Status.Commit
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecRecoverState(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	cmd := "cd " + HCL_DIR + ";" +
		"cat > terraform.tfstate << EOL\n" +
		apply.Status.State + "EOL"
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

func ExecTerraformDestroy(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	stdin io.Reader, stdout io.Writer, stderr io.Writer, apply *claimv1alpha1.TFApplyClaim) error {

	variables := GetTerraformVariables(apply)
	cmd := "cd " + HCL_DIR + ";" + "terraform destroy -auto-approve" + variables
	err := execPodCmd(client, config, podName, podNamespace, cmd, nil, stdout, stderr)

	if err != nil {
		return err
	}
	return nil
}

// execPodCmd exec command on specific pod and wait the command's output.
func execPodCmd(client kubernetes.Interface, config *restclient.Config, podName string, podNamespace string,
	command string, stdin io.Reader, stdout io.Writer, stderr io.Writer) error {

	cmd := []string{
		//"sh",
		"/bin/sh",
		"-c",
		command,
	}
	req := client.CoreV1().RESTClient().Post().Resource("pods").Name(podName).
		Namespace(podNamespace).SubResource("exec")

	option := &v1.PodExecOptions{
		Command: cmd,
		Stdin:   true,
		Stdout:  true,
		Stderr:  true,
		TTY:     true,
	}
	if stdin == nil {
		option.Stdin = false
	}
	req.VersionedParams(
		option,
		scheme.ParameterCodec,
	)

	exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
	if err != nil {
		return err
	}

	err = exec.Stream(remotecommand.StreamOptions{
		Stdin:  stdin,
		Stdout: stdout,
		Stderr: stderr,
	})
	if err != nil {
		return err
	}
	return nil
}

// ReadIDFromFile returns a Cloud Resource ID from Terraform State File
func ReadIDFromFile(filename string) (string, error) {
	var matched string // line with id
	var id string

	input, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Error(err, "Failed to read Terraform State File")
		return "", err
	}

	lines := strings.Split(string(input), "\n")

	for i, line := range lines {
		if strings.Contains(line, "\"id\"") {
			matched = lines[i]
		}
	}

	t := strings.Split(matched, "\"")

	if len(t) >= 4 {
		id = t[3]
	} else {
		err = errors.New("Index out of range")
		return "", err
	}

	return id, nil
}

// Copy a Dierectory (preserve directory structure)
func copy(source, destination string) error {
	var err error = filepath.Walk(source, func(path string, info os.FileInfo, err error) error {
		var relPath string = strings.Replace(path, source, "", 1)
		if relPath == "" {
			return nil
		}
		if info.IsDir() {
			return os.Mkdir(filepath.Join(destination, relPath), 0755)
		} else {
			var data, err1 = ioutil.ReadFile(filepath.Join(source, relPath))
			if err1 != nil {
				return err1
			}
			return ioutil.WriteFile(filepath.Join(destination, relPath), data, 0777)
		}
	})
	return err
}
